src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java

```diff
@@ -172,65 +177,95 @@ public Mono<T> first() {
+							: rs.query(statement, opts);
+				} else {
+					TransactionQueryOptions opts = buildTransactionOptions(pArgs.getOptions());
+					return (AttemptContextReactiveAccessor.createReactiveTransactionAttemptContext(s.get().getCore(),
+							clientFactory.getCluster().environment().jsonSerializer())).query(statement, opts);
+				}
+			});
+
+			return allResult.onErrorMap(throwable -> {
 				if (throwable instanceof RuntimeException) {
 					return template.potentiallyConvertRuntimeException((RuntimeException) throwable);
 				} else {
 					return throwable;
 				}
-			}).flatMapMany(ReactiveQueryResult::rowsAsObject).flatMap(row -> {
-				String id = null;
-				Long cas = null;
-				if (query.isDistinct() || distinctFields != null) {
-					id = "";
-					cas = Long.valueOf(0);
-				} else {
-					id = row.getString(TemplateUtils.SELECT_ID);
-					if (id == null) {
-						id = row.getString(TemplateUtils.SELECT_ID_3x);
-						row.removeKey(TemplateUtils.SELECT_ID_3x);
-					}
-					cas = row.getLong(TemplateUtils.SELECT_CAS);
-					if (cas == null) {
-						cas = row.getLong(TemplateUtils.SELECT_CAS_3x);
-						row.removeKey(TemplateUtils.SELECT_CAS_3x);
-					}
-					row.removeKey(TemplateUtils.SELECT_ID);
-					row.removeKey(TemplateUtils.SELECT_CAS);
-				}
-				return support.decodeEntity(id, row.toString(), cas, returnType, pArgs.getScope(), pArgs.getCollection());
-			}));
+			}).flatMapMany(o -> o instanceof ReactiveQueryResult ? ((ReactiveQueryResult) o).rowsAsObject()
+					: Flux.fromIterable(((TransactionQueryResult) o).rowsAsObject())).flatMap(row -> {
+						String id = "";
+						Long cas = Long.valueOf(0);
+						if (!query.isDistinct() && distinctFields == null) {
+							id = row.getString(TemplateUtils.SELECT_ID);
+							if (id == null) {
+								id = row.getString(TemplateUtils.SELECT_ID_3x);
+								row.removeKey(TemplateUtils.SELECT_ID_3x);
+							}
+							cas = row.getLong(TemplateUtils.SELECT_CAS);
+							if (cas == null) {
+								cas = row.getLong(TemplateUtils.SELECT_CAS_3x);
+								row.removeKey(TemplateUtils.SELECT_CAS_3x);
+							}
+							row.removeKey(TemplateUtils.SELECT_ID);
+							row.removeKey(TemplateUtils.SELECT_CAS);
+						}
+						return support.decodeEntity(id, row.toString(), cas, returnType, pArgs.getScope(), pArgs.getCollection(),
+								null, null);
+					});
 		}
 
 		public QueryOptions buildOptions(QueryOptions options) {
 			QueryScanConsistency qsc = scanConsistency != null ? scanConsistency : template.getConsistency();
 			return query.buildQueryOptions(options, qsc);
 		}
 
+		private TransactionQueryOptions buildTransactionOptions(QueryOptions options) {
+			TransactionQueryOptions opts = OptionsBuilder.buildTransactionQueryOptions(buildOptions(options));
+			return opts;
+		}
+
 		@Override
 		public Mono<Long> count() {
 			PseudoArgs<QueryOptions> pArgs = new PseudoArgs(template, scope, collection, options, domainType);
```

Comment:
Actually this seems to be in place in many methods, maybe doing the .defer methods should be in a follow-up changeset?